subsys/caf/modules/leds.c

```diff
@@ -0,0 +1,273 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <assert.h>
+#include <drivers/pwm.h>
+
+#include <caf/events/power_event.h>
+#include <caf/events/led_event.h>
+
+#include CONFIG_CAF_LEDS_DEF_PATH
+
+#define MODULE leds
+#include <caf/events/module_state_event.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_CAF_LEDS_LOG_LEVEL);
+
+#define LED_ID(led) ((led) - &leds[0])
+
+struct led {
+	const struct device *pwm_dev;
+
+	struct led_color color;
+	const struct led_effect *effect;
+	uint16_t effect_step;
+	uint16_t effect_substep;
+
+	struct k_delayed_work work;
+};
+
+static struct led leds[CONFIG_CAF_LEDS_COUNT];
+
+
+static void pwm_out(struct led *led, struct led_color *color)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(color->c); i++) {
+		int err = pwm_pin_set_usec(led->pwm_dev,
+					   led_pins[LED_ID(led)][i],
+					   CONFIG_CAF_LEDS_BRIGHTNESS_MAX,
+					   color->c[i],
+					   0);
+
+		if (err) {
+			LOG_ERR("Cannot set PWM output (err: %d)", err);
+		}
+	}
+}
+
+static void pwm_off(struct led *led)
+{
+	struct led_color nocolor = {0};
+
+	pwm_out(led, &nocolor);
+}
+
+static void work_handler(struct k_work *work)
+{
+	struct led *led = CONTAINER_OF(work, struct led, work);
+
+	const struct led_effect_step *effect_step =
+		&led->effect->steps[led->effect_step];
+
+	__ASSERT_NO_MSG(effect_step->substep_count > 0);
+	int substeps_left = effect_step->substep_count - led->effect_substep;
+
+	for (size_t i = 0; i < ARRAY_SIZE(led->color.c); i++) {
+		int diff = (effect_step->color.c[i] - led->color.c[i]) /
+			substeps_left;
+		led->color.c[i] += diff;
+	}
+	pwm_out(led, &led->color);
+
+	led->effect_substep++;
+	if (led->effect_substep == effect_step->substep_count) {
+		led->effect_substep = 0;
+		led->effect_step++;
+
+		if (led->effect_step == led->effect->step_count) {
+			if (led->effect->loop_forever) {
+				led->effect_step = 0;
+			} else {
+				struct led_ready_event *ready_event = new_led_ready_event();
+
+				ready_event->led_id = LED_ID(led);
+				ready_event->led_effect = led->effect;
+
+				EVENT_SUBMIT(ready_event);
+			}
+		}
+	}
+
+	if (led->effect_step < led->effect->step_count) {
+		int32_t next_delay =
+			led->effect->steps[led->effect_step].substep_time;
+
+		k_delayed_work_submit(&led->work, K_MSEC(next_delay));
+	}
+}
+
+static void led_update(struct led *led)
+{
+	k_delayed_work_cancel(&led->work);
+
+	led->effect_step = 0;
+	led->effect_substep = 0;
+
+	if (!led->effect) {
+		LOG_WRN("No effect set");
+		return;
+	}
+
+	__ASSERT_NO_MSG(led->effect->steps);
+
+	if (led->effect->step_count > 0) {
+		int32_t next_delay =
+			led->effect->steps[led->effect_step].substep_time;
+
+		k_delayed_work_submit(&led->work, K_MSEC(next_delay));
+	} else {
+		LOG_WRN("LED effect with no effect");
+	}
+}
+
+static int leds_init(void)
+{
+	static const char * const dev_name[] = {
+#if defined(CONFIG_PWM) && DT_NODE_HAS_STATUS(DT_NODELABEL(pwm0), okay)
```

Comment:
Maybe we could even assert that CONFIG_PWM is enabled (it's a Kconfig dependency of the module)

Reply:
Could add it but I don't think there is a need if there is a dependency in Kconfig.

Reply:
I leave this as is for now.